### ISSUE TITLE
GRPO 训练loss和reward异常

### ISSUE THREAD

**Issue by xiaosongyuan** [modelscope/ms-swift#3372]
**GRPO 做从 base model 开始的 RL 训练过程异常**
在 grpo 训练过程中不同指标出现异常：多个 rewards 为 0，kl 散度也为 0，如下图

![Image](https://github.com/user-attachments/assets/477428fb-3e86-4d1c-b065-00734e13a654)

completion 的采样结果全部为“！”，如下图

<img width="1127" alt="Image" src="https://github.com/user-attachments/assets/4a282914-3d82-47c6-aca3-ca7880baedcc" />

**训练脚本**
`sh requirements/install_all.sh
pip install transformers==4.46.3
pip install deepspeed==0.15.4
pip install math_verify
pip install git+https://github.com/huggingface/trl.git
pip install -e .

export MKL_THREADING_LAYER=pthreads
MASTER_PORT=29500 \
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=7 \
swift rlhf \
    --rlhf_type grpo \
    --model /mnt/models/models/public/Qwen/Qwen2.5/Qwen2.5-7B \
    --reward_funcs accuracy format \
    --use_vllm true \
    --vllm_device auto \
    --vllm_gpu_memory_utilization 0.7 \
    --vllm_max_model_len 8192 \
    --train_type full \
    --torch_dtype bfloat16 \
    --dataset custom_datasets/rl_data/openr1_sample_seed_42_grpo.jsonl \
    --max_completion_length 2048 \
    --num_train_epochs 2 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 2 \
    --eval_steps 40 \
    --save_steps 40 \
    --save_total_limit 3 \
    --save_only_model false \
    --logging_steps 1 \
    --max_length 4096 \
    --output_dir ckpt_rl/qwen2.5-7b/openr1_5k \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 8 \
    --dataset_num_proc 8 \
    --num_generations 7 \
    --temperature 0.6 \
    --system 'examples/train/grpo/prompt.txt' \
    --deepspeed zero2 \
    --log_completions true`

**Additional context**
尝试降低将默认的 temperature=0.9 降为 0.6，结果仍相同

**Comment by Jintao-Huang** [comment]
grad_norm nan了，导致训崩了

**Comment by xiaosongyuan** [comment]
rl 需要和 sft 一样设置 `--max_grad_norm 1.0`吗？

**Comment by xiaosongyuan** [comment]
刚试了设置传入 `--max_grad_norm 1.0` ，grad_norm 打印出来还是 `nan`，但是 reward/MathAccuracy 和 reward 值变正常了。

![Image](https://github.com/user-attachments/assets/189bef92-583b-4f56-a522-d365e12954e2)

**Comment by zerovl** [comment]
也遇到了这个问题，loss和kl都是0（beta设置了1e-3），命令行如下

`CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 MASTER_PORT=20202 NPROC_PER_NODE=8 MAX_PIXELS=301056 \
swift rlhf \
    --rlhf_type grpo \
    --model $MODEL_PATH \
    --external_plugins examples/train/grpo/plugin/plugin.py \
    --reward_funcs external_r1v_acc format \
    --use_vllm true \
    --train_type full \
    --torch_dtype bfloat16 \
    --dataset lmms-lab/multimodal-open-r1-8k-verified \
    --max_completion_length 1536 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 2 \
    --learning_rate 1e-6 \
    --eval_steps 1000 \
    --save_steps 1000 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --output_dir output \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --temperature 1.0 \
    --top_p 0.9 \
    --top_k 50 \
    --system 'examples/train/grpo/prompt.txt' \
    --vllm_gpu_memory_utilization 0.5 \
    --sleep_level 1 \
    --deepspeed zero3 \
    --num_infer_workers 8 \
    --tensor_parallel_size 4 \
    --log_completions true \
    --beta 1e-3 \
    --num_iterations 1 \
    --max_grad_norm 0.5 \
    --report_to wandb`

**Comment by hjh0119** [comment]
It seems to be a strange bug with vLLM under Zero2. I tried switching from zero2 to zero3 or using lmdeploy to resolve it. Maybe you could give it a try too.

**Comment by serenditipy-AC** [comment]
我也遇到了这种情况，GRPO+VLLM+Deepspeed zero3会出现梯度消失，loss为0的情况，但是切换到zero2会恢复，希望早日改进

**Comment by tastelikefeet** [comment]
> 刚刚更新了一下代码，我尝试了qwen2.5_VL 3B和7B的模型，进行的是视频微调，结果是一致的，是VLLM和zero3有冲突，梯度一直显示为0，如下图： ![Image](https://github.com/user-attachments/assets/4a8fb72e-e52c-4765-b592-38aae94a36b6)
> 
> 切换为zero2之后梯度是正常的，如下图：
> 
> ![Image](https://github.com/user-attachments/assets/7dbb184c-cf15-4e15-a29d-03ffd8fdec8a) 希望早日修复，辛苦啦，命令行代码如下 ` NPROC_PER_NODE=8 \ swift rlhf \ --rlhf_type grpo \ --model "/root/autodl-tmp/model/checkpoint-313" \ --train_type full \ --dataset "/root/autodl-tmp/data/grpo_trainset.jsonl" \ --torch_dtype bfloat16 \ --system "${SYSTEM_PROMPT}" \ --num_train_epochs 1 \ --reward_weights 0.3 0.7 \ --max_length 6144 \ --eval_steps 2000 \ --save_steps 100 \ --save_only_model true \ --target_modules all-linear \ --save_total_limit 1 \ --per_device_train_batch_size 1 \ --per_device_eval_batch_size 1 \ --learning_rate 1e-6 \ --logging_steps 1 \ --output_dir /root/autodl-tmp/port \ --warmup_ratio 0.05 \ --dataloader_num_workers 4 \ --max_completion_length 2048 \ --reward_funcs accuracy format \ --gradient_accumulation_steps 1 \ --use_vllm true \ --vllm_gpu_memory_utilization 0.4 \ --sleep_level 2 \ --tensor_parallel_size 4 \ --num_generations 8 \ --num_infer_workers 8 \ --temperature 1.0 \ --deepspeed zero3 \ --top_p 0.85 \ --log_level debug \ --log_level_replica warning \ 2>&1 | tee /root/autodl-tmp/grpo_7B_training.log`

试一下我的命令，我看了下是正常的：
```json
{
            "name": "grpo",
            "type": "debugpy",
            "request": "launch",
            "program": "swift/cli/main.py",
            "console": "integratedTerminal",
            "justMyCode": false,
            "args": [
                "rlhf",
                "--rlhf_type", "grpo",
                "--model", "Qwen/Qwen2.5-VL-7B-Instruct",
                "--reward_funcs", "accuracy", "format",
                "--use_vllm", "true",
                "--train_type", "lora",
                "--torch_dtype", "bfloat16",
                "--vllm_gpu_memory_utilization", "0.3",
                "--dataset", "AI-ModelScope/chartqa_digit_r1v_format",
                "--max_completion_length", "1024",
                "--num_train_epochs", "1",
                "--per_device_train_batch_size", "1",
                "--per_device_eval_batch_size", "1",
                "--learning_rate", "1e-6",
                "--gradient_accumulation_steps", "1",
                "--eval_strategy", "no",
                "--save_steps", "1000",
                "--save_total_limit", "2",
                "--logging_steps", "1",
                "--report_to", "wandb",
                "--max_length", "2048",
                "--output_dir", "output",
                "--warmup_ratio", "0.05",
                "--dataloader_num_workers", "4",
                "--dataset_num_proc", "4",
                "--num_generations", "8",
                "--temperature", "1.0",
                "--system", "examples/train/grpo/prompt.txt",
                "--deepspeed", "zero3_offload",
                "--num_infer_workers", "8",
                "--tensor_parallel_size", "4",
                "--log_completions", "true",
                "--async_generate", "false",
                "--sleep_level", "1",
            ],
            "env": {
                "NPROC_PER_NODE": "8",
                "nproc_per_node": "8",
                "PYTHONPATH": ".",
                "MASTER_PORT": "30000",
            }
        },
```
